### PR TITLE
Add List and delete capabilities to CF Scheduler Impl

### PR DESCRIPTION
**Pull request by cppwfs**

### CODE REVIEW

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundrySchedulerAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	public Scheduler scheduler(ReactorSchedulerClient client,
+			CloudFoundryOperations operations,
+			CloudFoundryConnectionProperties properties) {
+		return new CloudFoundryAppScheduler(client, operations, properties);
+	}
+
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConfigurationProperties(prefix = CloudFoundrySchedulerProperties.CLOUDFOUNDRY_PROPERTIES)
```
Comment (mminella) [inline src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundrySchedulerAutoConfiguration.java]: I think this mechanism shouldn't be used here.  See this in the documentation: https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-3rd-party-configuration

File: src/test/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppSchedulerTests.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+	private Applications applications;
+
+	@Mock(answer = Answers.RETURNS_SMART_NULLS)
+	private CloudFoundryOperations operations;
+
+	@Mock(answer = Answers.RETURNS_SMART_NULLS)
+	private Spaces spaces;
+
+	@Mock(answer = Answers.RETURNS_SMART_NULLS)
+	private ApplicationsV2 applicationsV2;
```
Comment (mminella) [inline src/test/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppSchedulerTests.java]: Why does this use V2?  V2 doesn't even know of the concept of a task...

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+		Assert.notNull(properties, "properties must not be null");
+
+		this.client = client;
+		this.operations = operations;
+		this.properties = properties;
+	}
+
+	@Override
+	public void schedule(ScheduleRequest scheduleRequest) {
+	}
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+					.build());
+		})
+				.onErrorMap(e -> {
+					if (e instanceof NoSuchElementException) {
+						throw new SchedulerException(String.format("Failed to unschedule, schedule %s does not exist.", scheduleName), e);
+					}
+					throw new SchedulerException("Failed to unschedule: " + scheduleName, e);
+				})
+				.cache()
+				.block();
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+								scheduleJobInfo.setScheduleProperties(new HashMap<>());
+								scheduleJobInfo.setScheduleName(job.getName());
+								scheduleJobInfo.setTaskDefinitionName(optionalApp.getName());
+								scheduleJobInfo.setJobId(job.getId());
+								return scheduleJobInfo;
+							});
+				});
+	}
+
+	private Flux<ScheduleInfo> getSchedulesList() {
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+	@Override
+	public List<ScheduleInfo> list() {
+		return getSchedulesList()
+				.collectList()
+				.cache()
+				.block();
+	}
+
+	private Flux<ScheduleJobInfo> getJobsList() {
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		return this.client.
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: Mistake on my part.  Put the '.' on the wrong line.

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+ *
+ * @author Glenn Renfro
+ */
+public class CloudFoundryAppScheduler implements Scheduler {
+
+	private SchedulerClient client;
+
+	private CloudFoundryOperations operations;
+
+	private CloudFoundryConnectionProperties properties;
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+	public Mono<ListJobSchedulesResponse> getScheduleExpression(String jobId) {
+		return this.client.jobs()
+				.listSchedules(
+						ListJobSchedulesRequest
+								.builder()
+								.jobId(jobId)
+								.build());
+	}
+
+	private Mono<ApplicationSummary> getApplication(String appId) {
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+				.filter(application -> appId.equals(application.getId()))
+				.singleOrEmpty();
+	}
+
+	private Flux<ApplicationSummary> requestListApplications() {
+		return this.operations.applications()
+				.list();
+	}
+
+	private String getSpaceId(String spaceName) {
```
[the remark this answers is not in the document]
Comment (cppwfs) [reply src/main/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppScheduler.java]: done

File: src/test/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppSchedulerTests.java (hunk 2 above)
Comment (cppwfs) [reply src/test/java/org/springframework/cloud/scheduler/spi/cloudfoundry/CloudFoundryAppSchedulerTests.java]: done

**Comment by cppwfs** [comment]
With the changes coming from the PCF-Scheduler Team and fixes to the Java Scheduler Client, this PR is no longer valid.  I will break up unschedule and base core work into a separate PR.

**Pull request closed without merging**